Mathlib/RingTheory/SimpleRing/Basic.lean

```diff
@@ -79,6 +80,37 @@ lemma isField_center (A : Type*) [Ring A] [IsSimpleRing A] : IsField (Subring.ce
       _ = y * (a * (x * y)) := by rw [mul_assoc a x y]
       _ = y * a := by rw [hy, mul_one]
 
+lemma injective_ringHom_or_subsingleton_codomain
+    {R S : Type*} [NonAssocRing R] [IsSimpleRing R] [NonAssocRing S]
+    (f : R →+* S) : Function.Injective f ∨ Subsingleton S :=
+  simple.eq_bot_or_eq_top (TwoSidedIdeal.ker f) |>.imp (TwoSidedIdeal.ker_eq_bot _ |>.1)
+    (fun h => subsingleton_iff_zero_eq_one.1 <| by
+      have mem : 1 ∈ TwoSidedIdeal.ker f := h.symm ▸ TwoSidedIdeal.mem_top _
+      rwa [TwoSidedIdeal.mem_ker, map_one, eq_comm] at mem)
+
+lemma injective_ringHom
+    {R S : Type*} [NonAssocRing R] [IsSimpleRing R] [NonAssocRing S] [Nontrivial S]
+    (f : R →+* S) : Function.Injective f :=
+  injective_ringHom_or_subsingleton_codomain f |>.resolve_right fun r => not_subsingleton _ r
```

Comment:
This could even replace the existing `RingHom.injective`, right?
See https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Field/Basic.html#RingHom.injective

If we put this in a file with minimal imports, I think that would be a good refactor.

Reply:
Indeed, let me try

Reply:
Wait, maybe no. Two-sided-ideal does not work on semiring, we really do need subtraction. So `TwoSidedIdeal.ker f` only makes sense for `f : R -> S` between `NonAssocRing`s. Therefore, `SimpleRing.injective_ringHom` need the target to have subtraction at all. `RingHom.injective` however works on target that is only a `Semiring`.

Reply:
I will write an implementation note to note this

Reply:
Ok, let's merge this as is now.
But I think we could generalize `TwoSidedIdeal` and `IsSimpleRing` so that they make sense in the context of non-assoc semirings.